File: rings.cpp
#include <iostream>
#include <map>
#include <set>
#include <string>

using namespace std;

// Функция для ввода строки с кольцами
string input() {
    string rings;
    cout << "Введите строку: ";
    cin >> rings;
    return rings;
}

// Функция для вывода результата
void output(int result) {
    cout << "Количество стержней с 3 цветами: " << result << endl;
}

int main() {
    string rings = input();
    // контейнер, где ключ - стержень, а значение - Множество (set) (хранить только уникальные значения) 
    map<int, set<char>> ma = {{0, {}}, {1, {}}, {2, {}}, {3, {}}, {4, {}},
    {5, {}}, {6, {}}, {7, {}}, {8, {}}, {9, {}}};

    // Проходим по строке с кольцами и цветами
    for (int i = 0; i < static_cast<int>(rings.size()); i += 2) {
        char color = rings[i]; // Цвет кольца
        int rod = rings[i + 1] - '0'; // Стержень кольца
        //- '0' — преобразует символ в число, так как rings[i + 1] возвращает символ, а не число
        ma[rod].insert(color); // Добавляем цвет в соответствующий стержень
    }

    int result = 0;// кол-во стрержней с 3 цветами

    // считаем содержимое массива
    for (int i = 0; i < 10; i++) {
        int c = 0;//счётчик цветов
        for (char n : ma[i]) {
            c++;
        }
        if (c == 3){//если на стержне все 3 цвета, добавляем 1 к результату
            result ++;
        }
    }

output(result);

return 0;
}

File: rings.rs
use std::collections::{HashMap, HashSet};
use std::io;

// Функция для ввода строки с кольцами
fn input() -> String {
    println!("Введите строку: ");
    let mut rings = String::new(); // Создание новой изменяемой строки для хранения ввода
    io::stdin()
        .read_line(&mut rings)
        .expect("Не удалось прочитать строку");
    rings.trim().to_string() // Удаление пробелов и переводов строки, возврат строки
}

// Функция для вывода результата
fn output(result: i32) {
    println!("Количество стержней с 3 цветами: {}", result);
}

fn main() {
    let rings = input();
    // Создаем HashMap, где ключ - стержень, значение - HashSet цветов
    let mut rods: HashMap<i32, HashSet<char>> = HashMap::new(); // i32 дефолтный целочисленный тип в Rust
    
    // Инициализируем стержни от 0 до 9
    for i in 0..10 {
        rods.insert(i, HashSet::new());
    }
    
    // Обрабатываем строку по два символа за раз
    let mut chars = rings.chars();
    while let (Some(color), Some(rod_char)) = (chars.next(), chars.next()) {
        // Чтение по два символа
        let rod = rod_char.to_digit(10).expect("Ожидалась цифра") as i32; // Преобразование символа в цифру
        rods.get_mut(&rod)
            .unwrap()
            .insert(color); // Получение изменяемой ссылки на стержень, извлечение значения, добавление цвета в множество
    }
    
    // Считаем стержни с 3 цветами
    let result = rods.values().filter(|colors| colors.len() == 3).count() as i32;
    output(result);
}